### PR TITLE
Accept bare word m4 macros

### PR DESCRIPTION
**Pull request by cgzones**
Accept m4 macros without any arguments.

Closes: #178

Downside:
Now we will accept all kind of artefacts.

### CODE REVIEW

File: src/parse.y [hunk 1]
```diff
@@ -702,8 +708,6 @@ m4_argument:
 	|
 	BACKTICK lines SINGLE_QUOTE
 	|
-	BACKTICK strings SINGLE_QUOTE { free_string_list($2); }
```
Comment (dburgener) [inline src/parse.y:705]: This seems concerning.  The commit message where this line was added mentions `define(`foo', `bar_t baz_t')` as something desirable to support.  My guess is that we'd now break on that?  It seems more desirable to me to support that then bare word macros.

Comment (dburgener) [reply src/parse.y:705]: Or would those parse under the lines, since they'd be interpreted as bare_word macros?

**Comment by dburgener** [comment]
Any idea why this worked before the multi-line change?  I don't remember the details of when I looked at it last, but I remember being somewhat confused as to what had changed.

**Comment by dburgener** [comment]
> Downside:
> Now we will accept all kind of artefacts.

I'm super torn on this.  Part of me says that throwing a parse error on a random stray character is highly desirable and part of me says that we can't force policy developers to avoid bare word macros or else not use selint.  The general goal should be to minimize breakage when the policy compiles.  @cgzones what are your thoughts on this tradeoff?  @stevedlawrence, do you have any opinions here?

**Comment by stevedlawrence** [comment]
Are bare word macros that common? I assume not since we would have since this before? If that's the case, what if we parse them as strings like this change does, but then add a "stray character check".  This way we error in the cases where a user really does have stray characters (which I agree seems pretty useful), but the check could still be disabled globally or with an "selint-disable" comment on lines where a user really does have a bare macro.

**Comment by dburgener** [comment]
That sounds like a good idea to me @stevedlawrence.  Yes, I think they're very uncommon.  I noticed this because of one in a policy I'm working on, but it's the only time I've seen it.  In that case, it was a pretty trivial fix to change it from "bare_word" to "bare_word()" and get it parsing, so it's not a huge deal if we don't support it.  What bothers me more is that the policy was parsing correctly and then I updated selint and it broke.

I like the general idea of parsing it and checking for stray words.  I'll see if I can get such a check written in time for 1.2.0.

File: src/parse.y (hunk 1 above)
Comment (cgzones) [reply src/parse.y:705]: Yes, they would parse as `lines`, which is `lines line` and `line` can be a `STRING`.

**Comment by cgzones** [comment]
I personally like SELint to be rather strict about unknown/uncommon constructs; to be warned early instead of finding out later a tool just hid suspicious code.

But I am fine moving this *strictness* from a hard parsing failure to a normal (ignorable) (error-)check.

**Comment by dburgener** [comment]
This all looks good to me.  I'm ready to merge it if you want to mark it as ready.

**Comment by dburgener** [comment]
@cgzones It looks like this broke this situation:

```
define(`not_foo',`-foo_t')
```

That's something we have in a policy I'm working on that I definitely need to support in my use case (moreso than the problem this PR addresses).

I tried adding "DASH STRING" into m4_simple_macro, but that shift/reduce conflicts with mls_range.  Any thoughts on a way to fix this problem?

My inclination is to say the supporting stray bare words may continue to cause problems in parsing and maybe the best call would be to revert this PR.  Thoughts?

**Comment by cgzones** [comment]
m4 arguments are quite hard to identify and parse.

Is #187 a sound solution?

**Comment by dburgener** [comment]
> m4 arguments are quite hard to identify and parse.

Very much so.  #187 seems reasonable at first blush, I think...  Let me take a little while to look/think more closely about it.

**Pull request merged**